containers/image.go

```diff
@@ -66,28 +57,97 @@ func DownloadImage(imgList []string, dest string) error {
 			log.Printf("Error when creating new image reference for %s", img)
 			return err
 		}
-		//////////
-
-		// Create systemContext to select os and arch
-		sysCtx := &types.SystemContext{
-			ArchitectureChoice: "amd64",
-			OSChoice:           "linux",
-		}
 
 		// Download and create tar
 		fmt.Printf("Copy %s to %s\n", img, dest)
-		_, err = copy.Image(context.Background(), policyContext, destRef, srcRef, &copy.Options{
-			ReportWriter: os.Stdout,
-			SourceCtx:    sysCtx,
-		})
+		err = copyImg(context.Background(), srcRef, destRef)
 		if err != nil {
 			log.Printf("Error when downloading image %s", img)
 			return err
 		}
+
+		// Fetching Cosing signature
+		// Do not fail at signature error (the signature may not exist yet)
+		if fetchSignature {
+			imgDigest, err := fetchDigest(srcRef)
+			if err != nil {
+				log.Printf("Error when fecthing digest image %s", srcRef.DockerReference().Name())
+				continue
+			}
+
+			// Creating Cosign reference
+			imgRefSig := fmt.Sprintf("%s%s:%s%s", "docker://", strings.Split(imgNameTagged.Name(), ":")[0], strings.Replace(imgDigest.String(), ":", "-", 1), ".sig")
+			srcRefSig, err := alltransports.ParseImageName(imgRefSig)
+			if err != nil {
+				log.Printf("Error when downloading signature %s", imgRefSig)
+				continue
+			}
+			// Creating directory for signature
+			destDirSig := fmt.Sprintf("%s/%s/%s%s", dest, "signature", strings.Replace(imgDigest.String(), ":", "-", 1), ".sig")
+			helpers.CreateDestDir(destDirSig)
+			path, _ := os.Getwd()
+			// Creating destination reference (Since it's not tar layered cannot reuse same method as above)
+			destRefSig, err := alltransports.ParseImageName(fmt.Sprintf("%s/%s/%s", "dir:", path, destDirSig))
+			if err != nil {
+				log.Printf("Error when downloading signature %s", imgRefSig)
+				continue
+			}
+			err = copyImg(context.Background(), srcRefSig, destRefSig)
+			if err != nil {
+				log.Printf("Error when downloading signature %s", imgRefSig)
+				continue
+			}
+		}
 	}
 	return nil
 }
 
+func copyImg(ctx context.Context, srcImgRef types.ImageReference, destImgRef types.ImageReference) error {
+	// Contexts
+	defaultPolicy, err := signature.NewPolicyFromFile("policy.json")
+	if err != nil {
+		log.Printf("Default policy err.\n")
+		return err
+	}
+	policyContext, err := signature.NewPolicyContext(defaultPolicy)
+	if err != nil {
+		log.Printf("Policy context err.\n")
+		return err
+	}
+	defer policyContext.Destroy()
+	// Create systemContext to select os and arch
+	sysCtx := &types.SystemContext{
+		ArchitectureChoice: "amd64",
+		OSChoice:           "linux",
+	}
+	_, err = copy.Image(context.Background(), policyContext, destImgRef, srcImgRef, &copy.Options{
+		ReportWriter: os.Stdout,
+		SourceCtx:    sysCtx,
+	})
+	return err
+}
+
+func fetchDigest(srcImgRef types.ImageReference) (*digest.Digest, error) {
```

Comment:
this function can be removed, the digest can be fetched from the return of copy.Image().